File: app/src/main/java/vn/loitp/app/activity/demo/ebookwithrealm/adapters/BooksAdapter.java
package vn.loitp.app.activity.demo.ebookwithrealm.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.core.utilities.LUIUtil;

import io.realm.Realm;
import vn.loitp.app.R;
import vn.loitp.app.activity.demo.ebookwithrealm.model.Book;
import vn.loitp.app.activity.demo.ebookwithrealm.realm.RealmController;

public class BooksAdapter extends RealmRecyclerViewAdapter<Book> {
    final Context context;
    private Realm realm;
    private LayoutInflater inflater;

    public interface OnClick {
        void onClick(Book book, int position);

        void onLongClick(int position);
    }

    private OnClick onClick;

    public BooksAdapter(Context context, OnClick onClick) {
        this.context = context;
        this.onClick = onClick;
    }

    @NonNull
    @Override
    public CardViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.real_item_books, parent, false);
        return new CardViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, final int position) {
        realm = RealmController.getInstance().getRealm();

        final Book book = getItem(position);
        final CardViewHolder holder = (CardViewHolder) viewHolder;

        holder.textTitle.setText(book.getTitle());
        holder.textAuthor.setText(book.getAuthor());
        holder.textDescription.setText(book.getDescription());

        holder.imageBackground.setBackgroundColor(LUIUtil.INSTANCE.getColor(context));
        if (book.getImageUrl() != null) {
            Glide.with(context)
                    .asBitmap()
                    .load(book.getImageUrl().replace("https", "http"))
                    //.fitCenter()
                    .into(holder.imageBackground);
        }

        //remove single match from realm
        holder.card.setOnLongClickListener(v -> {
            if (onClick != null) {
                onClick.onLongClick(position);
            }
            return false;
        });

        //update single match from realm
        holder.card.setOnClickListener(v -> {
            if (onClick != null) {
                onClick.onClick(book, position);
            }
        });
    }

    // return the size of your data set (invoked by the layout manager)
    public int getItemCount() {
        if (getRealmAdapter() != null) {
            return getRealmAdapter().getCount();
        }
        return 0;
    }

    public static class CardViewHolder extends RecyclerView.ViewHolder {
        CardView card;
        TextView textTitle;
        TextView textAuthor;
        TextView textDescription;
        ImageView imageBackground;

        CardViewHolder(View itemView) {
            super(itemView);
            card = itemView.findViewById(R.id.card_books);
            textTitle = itemView.findViewById(R.id.text_books_title);
            textAuthor = itemView.findViewById(R.id.text_books_author);
            textDescription = itemView.findViewById(R.id.text_books_description);
            imageBackground = itemView.findViewById(R.id.image_background);
        }
    }
}

File: app/src/main/java/vn/loitp/up/a/demo/ebookWithRealm/adt/BooksAdapter.kt
package vn.loitp.up.a.demo.ebookWithRealm.adt

import android.content.Context
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import vn.loitp.R
import vn.loitp.up.a.demo.ebookWithRealm.md.Book

class BooksAdapter(
    val context: Context,
    private val onClick: OnClick?
) : RealmRecyclerViewAdapter<Book?>() {

    interface OnClick {
        fun onClick(
            book: Book,
            position: Int
        )
        fun onLongClick(position: Int)
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): CardViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.v_item_books_realm, parent, false)
        return CardViewHolder(view)
    }

    override fun onBindViewHolder(
        viewHolder: RecyclerView.ViewHolder,
        position: Int
    ) {
        val book = getItem(position)

        book?.let { b ->
            if (viewHolder is CardViewHolder) {
                viewHolder.textBooksTitle.text = b.title
                viewHolder.textBooksAuthor.text = b.author
                viewHolder.textBooksDescription.text = b.description
                viewHolder.ivBackground.setBackgroundColor(Color.RED)

                Glide.with(context)
                    .asBitmap()
                    .load(b.imageUrl?.replace("https", "http")) // .fitCenter()
                    .into(viewHolder.ivBackground)

                // remove single match from realm
                viewHolder.cardBooks.setOnLongClickListener {
                    onClick?.onLongClick(position)
                    false
                }

                // update single match from realm
                viewHolder.cardBooks.setOnClickListener {
                    onClick?.onClick(b, position)
                }
            }
        }
    }

    // return the size of your data set (invoked by the layout manager)
    override fun getItemCount(): Int {
        return if (realmAdapter != null) {
            realmAdapter.count
        } else {
            0
        }
    }

    class CardViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var cardBooks: CardView = itemView.findViewById(R.id.cardBooks)
        var textBooksTitle: TextView = itemView.findViewById(R.id.textBooksTitle)
        var textBooksAuthor: TextView = itemView.findViewById(R.id.textBooksAuthor)
        var textBooksDescription: TextView = itemView.findViewById(R.id.textBooksDescription)
        var ivBackground: ImageView = itemView.findViewById(R.id.ivBackground)
    }
}
